File: 2025/Anna Park/Week4/2025-06-29-Search in Rotated Sorted Array.java
// First Solution
import java.util.*;

class Solution1 {
    public int search1(int[] nums, int target) {

        int[] sorted = new int[nums.length];
        int pivot = 0;

        for(int i=0; i<nums.length; i++){
            if (nums[i] < nums[pivot]) 
                pivot = i;
        }

        for (int i=0; i<nums.length; i++){
            sorted[i] = nums[(i+pivot)%nums.length];
        }

        int left = 0 , right = nums.length -1; 
        while (left<=right){
            int mid = (left + right)/2;
            if (target==sorted[mid])
                return mid+pivot;
            else if (target<=sorted[mid]){
                right = mid - 1;
            } else{
                left = mid + 1;
            }
        }

        return -1;
    }
}

//수정 후 코드
class Solution2 {
    public int search2(int[] nums, int target) {
        int left = 0, right = nums.length - 1;

        while (left <= right) {
            int mid = (left + right) / 2;

            if (nums[mid] == target)
                return mid;

            if (nums[left] <= nums[mid]) {
                if (nums[left] <= target && target < nums[mid])
                    right = mid - 1;
                else
                    left = mid + 1;

            } else {
                if (nums[mid] < target && target <= nums[right])
                    left = mid + 1;
                else
                    right = mid - 1;
            }
        }

        return -1;
    }
}

File: 2025/shinyoung/2025-06-29-search-in-rotated-sorted-array.py
def search(nums, target):
    n = len(nums)
    left = 0
    right = n - 1
    while left < right:
        middle = (left + right)//2
        if nums[middle] > nums[right]:
            left = middle+1
        else:
            right=middle
    min_index=left
    if min_index==0:
        left, right=0, n-1
    elif target >= nums[0] and target <= nums[min_index-1]:
        left, right = 0, min_index - 1
    else:
        left, right = min_index, n - 1
    
    while left <= right:
        middle = (left + right) // 2
        
        if nums[middle] == target:
            return middle
        elif nums[middle] < target:
            left=middle+1
        else:
            right=middle-1
    return -1
        


print(search([4, 5, 6, 7, 0, 1, 2], target = 0))
